07-13-16-language-plugins.md

```diff
@@ -0,0 +1,94 @@
+## A Milestone for Tessel
+This week we merged one of our most significant pull requests: [adding Rust deployment support](https://github.com/tessel/t2-cli/pull/790). For the past three years we've focused exclusively on deploying JavaScript on microcontrollers but this is our first step into a new language and a new community. We're not leaving JavaScript behind, rather, we see a future where the languages compliment each other (more on that in a subsequent post).
+
+Despite the significance of this addition to the CLI codebase, the amount of code flux was relatively small. We basically added two files: [one with the implementation](https://github.com/tessel/t2-cli/blob/b8cd1f34398c2d188fa188a216b4b9b4dd206dc4/lib/tessel/deployment/rust.js) and one with [unit tests](https://github.com/tessel/t2-cli/blob/b8cd1f34398c2d188fa188a216b4b9b4dd206dc4/test/unit/deployment/rust.js). That's a very good thing and it's largely due to the hardwork of [@rwaldron](https://github.com/rwaldron). In this post, I want to demonstrate how simple it is to add support for a new language to the CLI using the T2 language plugin system.
+
+The process for deploying a code bundle to Tessel can be generalized like so:
+
+1. Determine Project Location in FS ->
+2. Run any compilation/pre-processing steps ->
+3. Bundle executable code into tarball ->
+4. Run any pre-execution steps ->
+5. Execute the code/binary ->
+6. Run any post-execution steps ->
+
+As such, there are six functions that need to be exported for any language deployment mechanism:
+
+`checkConfiguration`, `preBundle`, `tarBundle`, `preRun`, `shell`/`binary`, and `postRun`.
+
+### checkConfiguration
+```.js
+// checkConfiguration to confirm project deployment files
+exportables.checkConfiguration = (pushdir, basename, program) => {
+  // Should return the basename and program name of project
+  return {
+    basename,
+    program
+  };
+};
+```
+
+`checkConfiguration` should be able to accept the directory that was deployed with `t2 run ...` and return the path to the entrypoint of the project. With Rust, we do this by parsing the `Cargo.toml` and with JavaScript we parse the project's `package.json`. It is expected to return an object with a `basename` and optional `program` property (for the name of the project).
+
+### preBundle
+```.js
+// Optional function to process options prior to bundling project
+exportables.preBundle = function(options) {
+  // Should return a Promise which eventually gets resolved
+  return Promise.resolve();
+};
+```
+`preBundle` is an optional step for any processes that need to be run prior to the next step where you actually package the directory up to be sent to the Tessel 2. It's less critical for Rust where we just set some options parameters but in JavaScript is integral to the binary compilation stage (where we inject our pre-compiled binaries in place of existing project binary dependencies).
```

Comment:
Can you give a more clear explanation for when you might need to use this step? I think the information is there, but it's tough to digest. Maybe just swap the order of talking about JS and Rust, and lead with "for example, in JS we need this"